doc: Extend esm examples, document and lint import forms for process + Buffer

guybedford opened the pull request:
This updates the use of esm examples in the documentation to cover more core modules while encouraging the use of Buffer and process imports / requires where possible, instead of saying that this is unnecessary. It does not deprecate or say that the global forms are deprecated though.

Summary:

* Split examples into ESM / CJS sections for: _async_context_, _async_hooks_, _cluster_, _dgram_, _diagnostics_channel_ (as far as I could go in one sitting!)
* Update both _process_ and _buffer_ to always show the import form in usage examples, noting this is encouraged but not required.
* In addition, destructured import forms are given preference where suitable.
* Add an eslint rule for the docs to not include process / Buffer without being imported and fix all cases.

@nodejs/modules

aduh95 commented on doc/api/async_hooks.md:
````diff
@@ -229,7 +306,15 @@ This behavior can be observed by doing something like opening a resource then
 
-```js
+```mjs
+import { createServer } from 'net';
+
+net.createServer().listen(function() { this.close(); });
````
```suggestion
import net from 'net';

net.createServer().listen(function() { this.close(); });
```
or
```suggestion
import { createServer } from 'net';

createServer().listen(function() { this.close(); });
```

aduh95 commented on doc/api/async_hooks.md:
````diff
@@ -270,12 +355,29 @@ created, while `triggerAsyncId` shows *why* a resource was created.
-const { fd } = process.stdout;
+```mjs
+import { createHook, executionASyncId } from 'async_hooks';
+import { stdout } from 'process';
````
```suggestion
import { stdout } from 'process';
import net from 'net';
```

aduh95 commented on doc/api/async_hooks.md:
````diff
@@ -270,12 +355,29 @@ created, while `triggerAsyncId` shows *why* a resource was created.
+
+```cjs
+const { createHook, executionAsyncId } = require('async_hooks');
+const { fd } = require('process').stdout;
````
```suggestion
const { fd } = require('process').stdout;
const net = require('net');
```

aduh95 commented on doc/api/async_hooks.md:
```diff
@@ -506,7 +608,17 @@ Using `executionAsyncResource()` in the top-level execution context will
+import { executionAsyncId, executionAsyncResource } from 'async_hooks';
+
+console.log(executionAsyncId(), executionAsyncResource());  // 1 {}
+open(fileURLToPath(import.meta.url), 'r', (err, fd) => {
```
```suggestion
open(new URL(import.meta.url), 'r', (err, fd) => {
```

bmeck commented on doc/api/async_context.md:
````diff
@@ -40,7 +44,39 @@ The following example uses `AsyncLocalStorage` to build a simple logger
-```js
+```mjs
+import http from 'http';
+import { AsyncLocalStoraget } from 'async_hooks';
````
```suggestion
import { AsyncLocalStorage } from 'async_hooks';
```

bmeck commented on doc/api/cluster.md:
````diff
@@ -387,7 +487,39 @@ added: v0.11.14
+```mjs
+import cluster from 'cluster';
+import http from 'http';
+import { cpus } from 'os';
````
import process from 'process';

bmeck commented on doc/api/cluster.md:
````diff
@@ -13,7 +13,37 @@ processes to handle the load.
+```mjs
+import cluster from 'cluster';
+import http from 'http';
+import { cpus } from 'os';
````
import process from 'process';

bmeck commented on doc/api/cluster.md:
````diff
@@ -204,7 +261,50 @@ See [`process` event: `'message'`][].
+```mjs
+import cluster from 'cluster';
+import http from 'http';
+import { cpus } from 'os';
````
```suggestion
import { cpus } from 'os';
import process from 'process';
```

bmeck commented on doc/api/dgram.md:
````diff
@@ -480,7 +542,17 @@ This method throws [`ERR_SOCKET_BAD_PORT`][] if called on an unbound socket.
 
-```js
+```mjs
+import dgram from 'dgram';
````
```suggestion
import dgram from 'dgram';
import { Buffer } from 'buffer';
```

bmeck commented on doc/api/dgram.md:
````diff
@@ -492,7 +564,18 @@ client.send(message, 41234, 'localhost', (err) => {
 
-```js
+```mjs
+import dgram from 'dgram';
````
```suggestion
import dgram from 'dgram';
import { Buffer } from 'buffer';
```

bmeck commented on doc/api/dgram.md:
````diff
@@ -510,7 +593,19 @@ however, sending multiple buffers is faster.
 
-```js
+```mjs
+import dgram from 'dgram';
````
```suggestion
import dgram from 'dgram';
import { Buffer } from 'buffer';
```

bmeck commented:
It might make sense to modify https://github.com/nodejs/node/blob/f4609bdf3fabdf441da6af17c2022565f4e18f9d/.eslintrc.js#L65-L101 and add a no-restricted globals for Buffer and process.

aduh95 commented on doc/api/crypto.md:
````diff
@@ -99,7 +99,7 @@ changes:
 
 ```mjs
-const { Certificate } = await import('crypto');
+import { Certificate } from 'crypto';
````
Note this was intentionally not using static import: https://github.com/nodejs/node/pull/37594#issuecomment-794021447. I think you should remove these changes from this PR, and open a new PR to discuss those if you think it should be changed.

guybedford replied:
Thanks for linking this to the original motivation @aduh95, I've reverted these. FWIW I disagree with the logic as there isn't a clear recovery path here. I would personally rather suggest always making the crypto module available, but providing a stub in non crypto builds that throws when attempting to use crypto to achieve the same effect. Agreed it's out of scope for this PR though.

guybedford commented:
Thanks, I've gone ahead and added the lint rule for `Buffer` and `process` and fixed up all cases.

I've also updated the process docs to use destructured import forms wherever possible as well eg `import { cwd } from 'process'`.

aduh95 commented on doc/api/async_hooks.md:
```diff
@@ -519,7 +633,33 @@ open(__filename, 'r', (err, fd) => {
+import {
+  executionAsyncId,
+  executionAsyncResource,
+  createHook
```
nit
```suggestion
  createHook,
```

aduh95 commented on doc/api/process.md:
```diff
@@ -941,12 +1208,24 @@ that exports a `foo` function. All the symbols are loaded before
 const module = { exports: {} };
-process.dlopen(module, path.join(__dirname, 'local.node'),
-               os.constants.dlopen.RTLD_NOW);
+dlopen(module, join(__dirname, 'local.node'), constants.dlopen.RTLD_NOW);
```
I subjectively prefer `path.join` rather than simply `join`, feel free to disregard if you disagree.
```suggestion
dlopen(module, path.join(__dirname, 'local.node'), constants.dlopen.RTLD_NOW);
```

guybedford replied:
I did go quite agressive on the destructured imports, down to leaving out cases like "on". I understand some of these get tricky, but want to try and encourage this as much as possible for tree-shaking style optimizations. In theory leaving out the member expression access is more performant as well although it's a micro-optimization for sure. Will leave this in unless anyone wants to push back further.

bmeck reviewed (approved):
LGTM, docs only change, would want to do some styling work to agree how to reduce noise viewing overall size of docs in a different PR.

guybedford commented:
Landed in https://github.com/nodejs/node/commit/f4d0a6a07bf72c556c5ba8e98b3b23327c273e80.

aduh95 commented:
@guybedford the lint rule only targets ESM syntax, is that on purpose? For info there are 135 occurrences of `Buffer` and `process` used as globals in CJS code snippets.
